File: src/lambda/dms-switch/utils/has-dms-changes.ts
import { CloudFormationClient, DescribeStacksCommand } from '@aws-sdk/client-cloudformation';
import { getChangeSet } from './get-change-set';

export const hasDmsChanges = async ({
  cf,
  StackName,
}: {
  cf: CloudFormationClient;
  StackName: string;
}): Promise<boolean> => {
  const describeCommand = new DescribeStacksCommand({ StackName });
  const stacks = await cf.send(describeCommand);
  console.log(JSON.stringify({ stacks }));
  const changes = await getChangeSet({
    cf,
    StackName,
    ChangeSetName: stacks.Stacks?.[0].ChangeSetId || '',
    changes: [],
  });
  const dmsChanges =
    changes.filter((change) => change.ResourceChange?.ResourceType?.startsWith('AWS::DMS')).length !== 0;
  return dmsChanges;
};

File: src/lambda/dms-switch-py/utils/has_dms_changes.py
import json
from utils.get_change_set import get_change_set


def has_dms_changes(cf, stack_name):
    stacks = cf.describe_stacks(StackName=stack_name)
    print(json.dumps({'stacks': stacks}, default=str))
    change_set_name = stacks['Stacks'][0]['ChangeSetId'] if stacks.get(
        'Stacks') else ''
    changes = get_change_set(cf, stack_name, change_set_name, [])
    dms_changes = any(change.get('ResourceChange', {}).get(
        'ResourceType', '').startswith('AWS::DMS') for change in changes)
    return dms_changes
